src/townname.cpp

```diff
@@ -107,28 +104,19 @@ char *GetTownName(char *buff, const Town *t, const char *last)
  */
 bool VerifyTownName(uint32 r, const TownNameParams *par, TownNames *town_names)
 {
-	/* reserve space for extra unicode character and terminating '\0' */
-	char buf1[(MAX_LENGTH_TOWN_NAME_CHARS + 1) * MAX_CHAR_LENGTH];
-	char buf2[(MAX_LENGTH_TOWN_NAME_CHARS + 1) * MAX_CHAR_LENGTH];
-
-	GetTownName(buf1, par, r, lastof(buf1));
+	std::string name = GetTownName(par, r);
 
 	/* Check size and width */
-	if (Utf8StringLength(buf1) >= MAX_LENGTH_TOWN_NAME_CHARS) return false;
+	if (Utf8StringLength(name) >= MAX_LENGTH_TOWN_NAME_CHARS) return false;
 
 	if (town_names != nullptr) {
-		if (town_names->find(buf1) != town_names->end()) return false;
-		town_names->insert(buf1);
+		if (town_names->find(name) != town_names->end()) return false;
+		town_names->insert(name);
 	} else {
 		for (const Town *t : Town::Iterate()) {
 			/* We can't just compare the numbers since
 			 * several numbers may map to a single name. */
-			const char *buf = t->name.empty() ? nullptr : t->name.c_str();
-			if (buf == nullptr) {
-				GetTownName(buf2, t, lastof(buf2));
-				buf = buf2;
-			}
-			if (strcmp(buf1, buf) == 0) return false;
+			if (name == (t->name.empty() ? GetTownName(t) : t->name)) return false;
```

Comment:
This condition feels confusing

Reply:
The alternative would be something like the following:
```suggestion
			if (t->name.empty()) {
				if (name == GetTownName(t)) return false;
			} else {
				if (name == t->name) return false;
			}
```
Although I could do a `std::string town_name = t->name.empty() ? GetTownName(t) : t->name;` and then compare on that. But that would copy the name, which seems kinda pointless to me.

So the main question is: is the suggestion better than the existing PR or not? And if that's not the case, how could it be made less confusing without the need to copy the town name?

Reply:
The suggestion seems fine to me. Could use std::string_view with the alternative?

Reply:
Be careful that we don't sit too much on the premature optimization; even in your suggestion, -O1 already doesn't make the copy anymore. If you put `const` in front of it, GCC already doesn't with -O0 (but clang and MSVC do). But if you make it a `const std::string &`, none of them make a copy :) But that ofc is just a `std::string_view` at that point :)

(sorry, I had some fun with godbolt :P I am always curious how much we have to keep those things in mind :))

Reply:
> The suggestion seems fine to me. Could use std::string_view with the alternative?

That would leave a dangling `std::string_view`, since `GetTownName` returns a `std::string` and that would go out of scope after the statement, i.e. after the view has been assigned. It will likely work fine most of the time, since the use is just after the release though.

Reply:
Just from an academical point of view, and as I like sharing things, I made two godbolts, one with:

`if (compare == (right.empty() ? GetTown() : right)) return 1;`

The other with:

```
        std::string test = right.empty() ? GetTown() : right;
        if (compare == test) return 1;
```

Did you know that the tenary operator already creates a new `std::string` in both cases? So both are nearly identical in result. Especially to how many `std::string`s are created :) So yeah, that solution would just have been fine :)

Your new solution works fine too ofc, just has 3 compares instead of one. Compilers are fun :D Right, enough nitpicking on a few lines of code :P